> 来源：SCNet 镜像 `o.o-llama-cpp-dcu:20260915` 内的 `/root/README.md`
> （md5 `4f10c238607fe6f4f398d3840792196f`，更新至 2026-09-14，与持久区 `README (7).md` 为同一份）。
> 以下为**原文照录、未作删改**；文中出现的本机绝对路径（`/root/...`、`/tmp/...`）不保证存在于其他实例。

# llama.cpp on Hygon DCU K100_AI (gfx928) — 修复报告

> 更新：2026-09-13
> 结论：**量化模型已修复并验证可用**（Q4_0 端到端正常出字）。历史排障过程另见
> `/root/CodeBuddy/20260912000526/README.md`。
> **新卡实测见 §11**：BW(gfx936) 上源码一行不用改，重新编译即可，但量化 + HIP graph 会崩（已定性）。
> **§12（2026-09-14）**：z100sm(gfx906) 实测通过，源码同样零改动；但镜像里现成的 `build-gfx9-all`
> 在这台 C86 7185 主机上会 **SIGILL（AVX-512 不匹配）**，必须按 §12.3 重编一份。
> **§13（2026-09-14）**：回到 K100_AI(gfx928) 真机复测 —— `build-gfx9-all2`（5 架构 fatbin）
> 拿来就能跑，全量 `MUL_MAT` **1563 / 0 FAIL**；同时**更正 §11.6 的运行期参数结论**
> （UM/SDMA 不需要、HIP graph 可用、`--gpu-max-threads-per-block` 是无效参数）。
> **§14（2026-09-14，最新）**：BW100(gfx936) 用同一份 `build-gfx9-all2` 上机复测（此前只编译未上机）——
> 全量 `MUL_MAT` **1193 / 0 FAIL**，端到端正常；两处更正：**F16 开 graph 也崩**（§11.4 记反了）、
> **FA 开关在噪声内**（§11.6 的"tg 慢 15~25%"不复现）。

---

## 0. TL;DR

| 项 | 结论 |
|---|---|
| 问题 | 量化 GGUF（Q4_0/Q4_K/IQ* …）在这张卡上输出满屏 `??????`，F16/BF16 正常 |
| 根因 | **host 与 device 对 gfx928 的架构判定不一致**，导致 MMQ/MMVQ 的 tile 配置与共享内存布局错位 |
| 关键约束 | gfx928 **没有可用的 MFMA**：DTK 26.04 的 clang 未开启 `mai-insts`，MFMA 内建函数编译失败 |
| 修复 | device 端补 `CDNA2`（对齐配置表）+ 屏蔽 MFMA 路径 + 还原作者的 mmq warp hack |
| 状态 | 全量 `MUL_MAT` 1195 项 0 FAIL；Q4_0 端到端输出正常，`Prompt 194.6 t/s / Gen 94.9 t/s` |
| 构建 | `/root/llama.cpp/build-gfx928/bin/`（旧的 `build/` 已于 2026-09-12 删除） |
| 「F16 prompt 只有 3 t/s」 | **是测量假象**，真实 `pp512 = 701 t/s`，见 §4 |

---

## 1. 环境

```
卡        : Hygon DCU K100_AI, gfx928:sramecc+:xnack-, 64GB (65520 MiB), wave64, 128 CU
驱动      : hydcu 6.2.28
DTK       : /opt/dtk -> DTK 26.04, dcc 25.10.0, clang 17.0.0
llama.cpp : /root/llama.cpp, commit 7584430 (0.2.0-dev)
主机      : CentOS 7 内核 + 容器内 Ubuntu 22.04.5；Hygon C86 7375 32c/64t, ~1TB 内存
```

`hy-smi` / `rocminfo` 的「驱动未加载」是**假报警**，镜像里缺 `lsmod`：

```bash
printf '#!/bin/sh\ncat /proc/modules\n' > /usr/local/bin/lsmod && chmod +x /usr/local/bin/lsmod
hy-smi && rocminfo | grep -E "gfx|Marketing"
```

---

## 2. 根因分析

### 2.1 主因：host / device 取到两套不匹配的配置

`ggml_cuda_parse_id("gfx928")` → `cc = 0x1000928`，落在 CDNA2 区间 →
**host 侧**按 CDNA 处理；而 `ggml/src/ggml-cuda/vendors/hip.h` 里只有
`__gfx908__/__gfx90a__/__gfx942__/__gfx950__` 等宏，**gfx928 一个都不匹配** →
**device 侧**落到 `#else` 的 RDNA2 兜底。

| 判定点 | host（CDNA） | device（RDNA2 兜底） |
|---|---|---|
| MMQ 配置表 | `mmq-config-cdna.cuh`：nthreads **512** | `mmq-config-rdna2.cuh`：nthreads **256** |
| 共享内存布局 | `use_mma_data_layout(cc) = true`（MMA 布局） | `use_mma_data_layout() = false`（dp4a 布局） |
| MMVQ 参数表 | `MMVQ_PARAMETERS_GCN`：ncols_dst=1 → **2 warps** | `MMVQ_PARAMETERS_GENERIC` → **4 warps** |

host 按 CDNA 表算 shmem 大小与网格维度，内核却按 RDNA2 表的 `tile_x`/偏移填 tile →
共享内存越界/错位 → GPU 上量化 `MUL_MAT` 出 NaN → 乱码。

### 2.2 次因：作者对 MMQ 的 hack

`ggml/src/ggml-cuda/mmq.cuh:1399` 被改成 `nwarps > 2 ? 2 : nwarps`（把 warp 数压到最多 2），
而内核里的 `nwarps` 是编译期从 config 推导的 → launch 与内核假设不符。已还原为上游写法。

### 2.3 关键发现：gfx928 不能走 MFMA

只补 `CDNA2` 会连带把 `AMD_MFMA_AVAILABLE` 打开，编译直接失败：

```
mma.cuh:1250:20: error: '__builtin_amdgcn_mfma_f32_16x16x16f16' needs target feature mai-insts
```

符号存在 ≠ 能编译：DTK 26.04 的 clang 没有为 gfx926/928/936/938 开启 `mai-insts`。
所以 gfx928 是 **「CDNA 的 MMQ 配置表 + 没有 MFMA」** 的混合体，任何"一刀切当 CDNA2"
或"一刀切当 RDNA2"的做法都不对。

---

## 3. 修复内容

共 3 处改动（完整状态补丁：`/root/CodeBuddy/20260912113812/final-working-state.patch`）：

**① `ggml/src/ggml-cuda/vendors/hip.h`** —— 在 CDNA 判定块前补：

```c
#if defined(__gfx926__) || defined(__gfx928__) || defined(__gfx936__) || defined(__gfx938__)
#define CDNA2
#define HYGON_DCU_NO_MFMA   // 这些目标没有 mai-insts，MFMA 编不过
#endif
```

**② `ggml/src/ggml-cuda/common.cuh`**

- `AMD_MFMA_AVAILABLE` 增加 `&& !defined(HYGON_DCU_NO_MFMA)`；
- 新增 `GGML_CUDA_CC_HYGON_GFX9{26,28,36,38}` 与 `GGML_CUDA_CC_IS_HYGON_DCU(cc)`；
- `amd_mfma_available(cc)` 对 Hygon 返回 `false`。

> host 侧拿不到 `__gfx928__` 这类 device 宏，所以必须额外有一套 cc 判定，两边保持同步。

**③ `ggml/src/ggml-cuda/mmq.cuh:1399`** —— 还原 `const dim3 block_dims(warp_size, nwarps, 1);`
（还原后该文件与上游一致，补丁里看不到这一项）

> 改完**不需要任何额外 CMake 开关**：默认的 `GGML_HIP_MMQ_MFMA=ON` 也能编过、也能跑对。

### 编译

```bash
cd /root/llama.cpp
source /opt/dtk/env.sh
cmake -B build-gfx928 -S . \
  -DCMAKE_BUILD_TYPE=Release \
  -DGGML_HIP=ON -DAMDGPU_TARGETS=gfx928 -DGPU_TARGETS=gfx928 \
  -DGGML_HIP_GRAPHS=ON -DGGML_HIP_NO_VMM=ON \
  -DCMAKE_HIP_COMPILER=/opt/dtk/llvm/bin/clang
cmake --build build-gfx928 -j64 --target llama-cli test-backend-ops llama-bench
```

- 注意：**不要**加 `-DLLAMA_BUILD_SERVER=OFF`，`llama-cli` 依赖 `llama-server-impl`。
- 注意：`ggml-hip/CMakeLists.txt` 里被删掉的「至少 ROCm/HIP 6.1」版本检查**必须保持删除**（DTK 报的版本过不了该检查）。

---

## 4. 验证证据

### 4.1 逐算子精度（`test-backend-ops`）

| 测试 | 修复前 | 修复后 |
|---|---|---|
| `-o MUL_MAT -p q4_0` | NaN / ERR，FAIL | **46/46 OK** |
| `-o MUL_MAT -p q4_K` | FAIL | **43/43 OK** |
| `-o MUL_MAT -p "iq"`（iq1_s/iq2_*/iq3_*/iq4_*） | FAIL | **132/132 OK** |
| `-o MUL_MAT`（全量） | 大量 FAIL | **1195 OK / 0 FAIL**，`2/2 backends passed` |

```bash
cd /root/llama.cpp/build-gfx928/bin && source /opt/dtk/env.sh && export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
./test-backend-ops test -o MUL_MAT -p "q4_0"
```

### 4.2 端到端生成

```bash
./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -p "你好" -n 96 -t 8 -st --no-warmup < /dev/null
```

修复前：`????????????????????????????????`（乱码）
修复后：

```
> 你好
[Start thinking]
Thinking Process:
1.  **Analyze the Input:**
    *   Input: "你好" (Nǐ hǎo)
    *   Language: Chinese
    *   Meaning: "Hello" / "Hi"
    *   Intent: Greeting, starting a conversation.

2.  **Determine the Appropriate Response:**
    *   Acknowledge the greeting.
    *   Maintain a friendly and helpful tone.

[ Prompt: 194.6 t/s | Generation: 94.9 t/s ]
```

### 4.3 `llama-bench`（512 token 真实 prefill）

| 模型 | 后端 | pp512 | tg32 |
|---|---|---|---|
| Qwen3.5-4B **F16** | ROCm `-ngl 99` | **701.53 ± 10.75 t/s** | **60.37 ± 0.19 t/s** |
| Qwen3.5-4B **Q4_0** | ROCm `-ngl 99` | **740.53 ± 8.89 t/s** | （见 §6 待复测） |

### 4.4 更正：不存在「F16 prompt 只有 3 t/s」这个问题

早期记录里的 `Prompt 2.9–3.0 t/s` 是**测量假象**：用 3 个 token 的 prompt 加
`--no-warmup`，把首次 kernel 加载 / 首次 GEMM 的一次性开销摊到 3 个 token 上。
真实数据是 `pp512 = 701 t/s`（F16）、`740 t/s`（Q4_0）。

也确认了**不是 CPU 在跑**：

```
load_tensors: offloaded 33/33 layers to GPU
load_tensors:  ROCm0 model buffer size = 8023.67 MiB
（CPU 层数 = 0；CPU_Mapped 1212 MiB 只是 mmap 文件映射，不参与计算）
```

### 4.5 仍然存在的（轻微）性能短板

- **MMQ 只支持量化类型**（`mmq.cu:299`，F16/BF16 `mmq_supported=false`），
  所以 F16 的 prefill matmul 只能走 **hipBLAS**。
- DTK 的 rocBLAS Tensile 内核库里**只有 gfx906**（`/opt/dtk/lib/rocblas/library` 237 个文件，
  架构标记全是 gfx906），没有 gfx928 的调优版本 → prefill 只能跑非调优 kernel
  （701 t/s ≈ 6 TFLOPS，远低于 K100 峰值）。
- 但 `/opt/dtk/lib/hipblaslt/library*` 里**有 gfx926/928/936/938 的内核** →
  可试 `ROCBLAS_USE_HIPBLASLT=1` 看 prefill 能否再快。
- `MMQ_DP4A_MAX_BATCH_SIZE = 64`（`mmq.cuh:8`）：量化模型 prompt **> 64 token** 时也会
  离开 MMQ 去走 hipBLAS。`GGML_CUDA_FORCE_MMQ` 可强制走 MMQ，值得对比测一轮。

---

## 5. 日常使用速查

```bash
# 每次新 shell
source /opt/dtk/env.sh
cd /root/llama.cpp/build-gfx928/bin
export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
```

**多轮对话**（不要加 `-p`）：

```bash
./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -t 8
```

> 会话内命令：`/exit`（或 Ctrl+C）、`/regen`、`/clear`、`/read <文件>`、`/glob <通配>`

**单轮问答**：

```bash
./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -t 8 -p "你好" -n 256 -st < /dev/null
```

常用参数：`-ngl 99`（全部层上 GPU，必须）、`-n 256`（生成上限）、`-c 4096`（上下文）、
`-sys "..."`（系统提示）、`--temp 0.7`、`--seed 42`、`-fa on`、`-mli`（多行输入）、
`-co off`（重定向到文件时关彩色）、`--mm /root/mmproj-BF16.gguf`（多模态）。

模型：

| 文件 | 大小 | 用途 |
|---|---|---|
| `/root/Qwen3.5-4B-Q4_0.gguf` | 2.5G | 日常首选，量化，已验证正常 |
| `/root/qwen3.5-4b-f16.gguf` | 7.9G | 全精度，验证时由 Q4_0 升转而来 |
| `/root/mmproj-BF16.gguf` | 645M | 多模态投影 |

---

## 6. 现场状态与遗留问题

### 6.1 卡死的测试进程（**已解决 ✅**）

**2026-09-12 15:00 解决**：执行 `pkill -9` 后 `hy-smi` 的 VRAM 从 62% 回到 **0%**，GPU 立刻恢复可用，
§10 的三条测试一次全过（27B Q2_K 在 GPU 上 28.7 t/s）。下次遇到同样情况照抄这两行：

```bash
pkill -9 -f llama-bench; pkill -9 -f "llama-cli -m /root"
sleep 2; source /opt/dtk/env.sh; export PATH=/usr/local/bin:$PATH; hy-smi   # 期望 VRAM% → 0
```

若 `pkill` 后 VRAM 不降 → KFD 队列已死，重启容器（同一个容器重启，`build-gfx928` 会保留）。

以下为当时的现象记录（供以后排查参考）。

并发跑测试时留下了几个卡在 GPU 事件上等待的进程（`kfd_wait_on_events`，
`hy-smi` 显示 `DCU% = 0`、VRAM 被占住），新进程会卡在 `Loading model...`：

```bash
pkill -f llama-bench; pkill -f "llama-cli -m /root"; pkill -f "gdb -batch"
sleep 2; source /opt/dtk/env.sh; export PATH=/usr/local/bin:$PATH; hy-smi   # 看 VRAM% 是否回落
```

若清理后新进程仍卡住 / VRAM 不释放 → KFD 队列已卡死，**重启容器**即可
（重启后记得重做 §1 的 `lsmod` shim）。

### 6.2 未定性：MMVQ 的 launch bounds 警告 + segfault

在卡死环境下跑 `llama-bench -m Q4_0 -p 512 -n 32 -ngl 99` 时，tg 阶段出现：

```
Launch params (256, 1, 1) are larger than launch bounds (128) for kernel mul_mat_vec_q<...>
Segmentation fault (core dumped)
```

- 256 = 4 warps（GENERIC 表）、128 = 2 warps（GCN 表），与 §2.1 的 host/device 取表差异同源。
- **但**同一构建下 `llama-cli` 用 Q4_0 生成 96 个 token **完全正常**，全量 `MUL_MAT`
  1195 项也全过，所以**不能断定是真 bug**，也可能是 GPU 已卡死后的连锁反应。
- 待办：环境清理干净后重跑一次定性；若稳定复现，查 `mmvq.cu` 的
  `calc_nwarps`（host 用 `get_device_table_id(cc)`、device 用 `get_device_table_id()`）
  是否还有别的分叉。

### 6.3 其它待办

- 试用 `ROCBLAS_USE_HIPBLASLT=1` / `GGML_CUDA_FORCE_MMQ` 对比 prefill 性能（§4.5）。
- 作者那批 `1024/512 → 256` 的 block 大小改动**保留未还原**：本次验证全过说明它们无害，
  但"是否必要"没单独验证。顺带一提 `GGML_CUDA_MAX_THREADS_PER_BLOCK`
  是个**没有任何引用的死定义**（`common.cuh:2`、`norm.cu:2`），可删。
- 未用 `hipprof` 确认量化路径实际跑的 kernel 名。

---

## 7. 关键路径

| 路径 | 说明 |
|---|---|
| `/root/llama.cpp/build-gfx928/bin/` | **可用构建**（llama-cli / test-backend-ops / llama-bench） |
| `/root/llama.cpp/build-gfx906/bin/` | z100sm 构建，2026-09-14 实测通过（§12） |
| `/root/llama.cpp/build-gfx9-all2/bin/` | **5 架构 fatbin**（906;926;928;936;938），跨主机可跑（§12.10） |
| ~~`/root/llama.cpp/build-gfx9-all/`~~ | 3 架构，但是 AVX-512 二进制，只在 C86 7375 主机上能用 |
| ~~`/root/llama.cpp/build/bin/`~~ | 旧构建（量化会乱码），**已于 2026-09-12 删除** |
| `/root/llama.cpp/ggml/src/ggml-cuda/vendors/hip.h` | 改动 ①：CDNA2 + HYGON_DCU_NO_MFMA |
| `/root/llama.cpp/ggml/src/ggml-cuda/common.cuh` | 改动 ②：AMD_MFMA_AVAILABLE / amd_mfma_available |
| `/root/llama.cpp/ggml/src/ggml-cuda/mmq.cuh:1399` | 改动 ③：还原 warp hack |
| `/root/CodeBuddy/20260912113812/final-working-state.patch` | 当前可用状态补丁 |
| `/root/CodeBuddy/20260912113812/author-changes.patch` | 动手前的作者原始改动备份 |
| `/root/CodeBuddy/20260912000526/README.md` | 历史排障过程记录（§4/§6 的旧结论已被本文更正） |
| `/opt/dtk/bin/hipprof`, `/opt/hyhal/bin/hy-smi` | profiling / 设备状态 |
| `/usr/local/bin/lsmod` | 本机新增 shim，让 hy-smi/rocminfo 正常工作 |

---

## 8. 移交给 ComfyUI 镜像：合并说明

> 给 ComfyUI 镜像里那个 CodeBuddy（或接手的人）看的操作手册。

### 8.1 先分清哪些能带走、哪些不能

| 类别 | 内容 |
|---|---|
| ✅ 可移植 | 根因分析（§2）、修复补丁、验证方法（§4）、各类坑（§3/§8.6） |
| ❌ 别照抄 | 本容器的绝对路径（`/root/llama.cpp`、`/root/*.gguf`）、`/usr/local/bin/lsmod` shim、DTK 版本号、`/dev/dri` 映射 |
| ⚠ 依赖版本 | 补丁基于 llama.cpp **commit 7584430**；换版本先确认 `common.cuh` / `vendors/hip.h` 结构是否一致（`git apply --check` 会告诉你） |

### 8.2 交接物清单

| 文件 | 说明 |
|---|---|
| `/root/hygon-gfx928-mmq-fix.patch` | **最小修复补丁**（3 个 hunk，只含本次修复，已校验可干净 `patch -p1`，结果与修复版逐字节一致） |
| `/root/README.md` | 本报告 |
| `/root/CodeBuddy/20260912113812/final-working-state.patch` | 本机完整工作状态（含作者遗留改动，仅作对照，**不要**直接拿去合并） |
| `/root/CodeBuddy/20260912113812/author-changes.patch` | 动手前的作者原始改动备份（回滚用） |

> 这些文件都在**当前容器**里。ComfyUI 是另一个镜像 → 先把它们带出去（复制文件到共享目录，或直接把文本贴过去）。

### 8.3 新环境第一步：先判断「要不要修」，别盲抄

```bash
# 1) 架构
rocminfo | grep -E "gfx|Marketing"                    # 期望 gfx926/928/936/938 之类

# 2) 设备识别：看 (0xNNN) 与 Wave Size
./llama-cli -m <任一 gguf> -ngl 0 -n 1 -p x 2>&1 | grep -i "Wave Size"

# 3) 是否同样乱码（这台的症状）
./test-backend-ops test -o MUL_MAT -p "q4_0"

# 4) 该 arch 有没有 mai-insts（决定能不能用 MFMA）
printf '__global__ void k(){ __builtin_amdgcn_mfma_f32_16x16x16f16(x,x,x,0,0,0); }\n' > /tmp/m.cu
clang --offload-arch=gfx928 -c /tmp/m.cu -o /tmp/m.o   # 报 needs target feature mai-insts → 无 MFMA

# 5) hipBLAS 内核覆盖（只关心性能时看）
ls /opt/dtk/lib/rocblas/library | grep -oE 'gfx[0-9a-z]+' | sort -u
```

判定表：

| 现场情况 | 该怎么处理 |
|---|---|
| 量化乱码 + host 认成 CDNA + device 无对应宏 | **就是本报告的问题** → 上补丁 |
| 编译报 `needs target feature mai-insts` | 该 arch 没有 MFMA → 必须同时屏蔽（补丁已含 `HYGON_DCU_NO_MFMA`） |
| `mai-insts` 可用（比如更新的 DTK） | **不要**照抄 `HYGON_DCU_NO_MFMA`，保留 MFMA 性能更好；补丁里只留 `CDNA2` 那几行 |
| 架构不是 0x92x / 0x93x | 改补丁里的宏列表与 cc 常量，且 `hip.h` 与 `common.cuh` **两处必须同步** |

### 8.4 合并步骤

```bash
cd <ComfyUI 镜像里的 llama.cpp>
git apply --check /path/to/hygon-gfx928-mmq-fix.patch    # 先干跑
patch -p1 < /path/to/hygon-gfx928-mmq-fix.patch          # 或 git apply

# 如果对方源码里保留了作者那个 mmq hack，还要手工还原：
grep -rn "nwarps > 2 ? 2 : nwarps" ggml/src/ggml-cuda/mmq.cuh   # 改为 nwarps
```

编译命令与两个坑见 §3（**不要**关 `LLAMA_BUILD_SERVER`；`ggml-hip/CMakeLists.txt` 的 HIP 版本检查按需删除）。

### 8.5 合并后必做的验证（三条，缺一不可）

```bash
./test-backend-ops test -o MUL_MAT -p "q4_0"    # 期望 46/46 OK（修复前是 NaN/FAIL）
./test-backend-ops test -o MUL_MAT             # 期望 1195 OK / 0 FAIL
./llama-cli -m <量化模型> -ngl 99 -p "你好" -n 64 -st --no-warmup < /dev/null   # 期望正常中文
```

> 提醒：**不要**用「几个 token 的短 prompt + `--no-warmup`」来判断快慢（§4.4 就是被这个坑过）。

### 8.6 ComfyUI 侧额外注意事项

- ComfyUI 常通过 `llama-cpp-python` / 自带 GGUF 节点加载模型：**先确认实际加载的是哪一份 llama.cpp**
  （`python -c "import llama_cpp; print(llama_cpp.__file__)"`、`ldd` 看 `.so` 来源）。
  本补丁只适用于**直接编译 llama.cpp** 的那一份。
- 若走 `llama-cpp-python`：它自带 `vendor/llama.cpp` 子模块，需要在**它的**那份源码上做同样处理，
  然后重装（`pip install -e .` 或 `CMAKE_ARGS=... pip install --no-cache-dir -e .`）。
- 别在同一张卡上同时跑 ComfyUI 推理和 llama.cpp 压测：显存/队列会互相拖死
  （本机就是这么把 GPU 搞卡的，见 §6.1）。
- ComfyUI 镜像的 DTK/驱动版本若不同，§4.5 的 hipBLAS 结论要重测。

---

## 9. 附：版本与升级（2026-09-12 调研结论）

| 项 | 本机（当前在用） | 上游 master |
|---|---|---|
| commit | `7584430` | `c069aa7` |
| 日期 | 2026-08-24 | 2026-09-12 |
| 版本 | ggml `0.21.0` / llama.cpp `0.2.0` | ggml `0.23` |
| 差距 | —— | 落后 **312** 个提交 |

**能不能升级：能，但只能走 tarball。**

- 本机仓库是**单提交压扁快照**（`git rev-list --count HEAD` = 1），且 `git` 协议在代理下不通
  （`git ls-remote` 空返回）→ **不能 `git pull`**，只能重新拿源码。
- `codeload.github.com` 可用（已验证 HTTP 200）：

  ```bash
  curl -L -o llama.cpp-master.tar.gz https://codeload.github.com/ggml-org/llama.cpp/tar.gz/refs/heads/master
  mkdir -p /root/llama.cpp-new && tar xzf llama.cpp-master.tar.gz -C /root/llama.cpp-new --strip-components=1
  cd /root/llama.cpp-new && patch -p1 --dry-run < /root/hygon-gfx928-mmq-fix.patch
  ```

**已预先验证完的事实**（下次不用再查）：

- `/root/llama.cpp-master.tar.gz`（35.6MB，即当天 master）已下载，并解压到 `/root/llama.cpp-new`。
- **补丁在最新 master 上 3 个 hunk 全部干净命中**（行号偏移 3 / 10 / 10）→ 兼容 ggml 0.23。
- **上游至今没有 gfx928 支持**：`vendors/hip.h` 里仍然只有 `__gfx908__ / __gfx90a__ / __gfx942__ / __gfx950__`；
  整个 `ggml-cuda/` 搜不到 `Hygon` / `gfx928` / `0x928`；`mmq.cu` 的
  `(!GGML_CUDA_CC_IS_CDNA(cc)) || ne11 < MMQ_DP4A_MAX_BATCH_SIZE` 规则也没变。
  → **补丁仍然必要，而且不用改就能用在新版上。**
- 升级收益只有两个：① 支持更新的模型架构；② 吃上游 HIP/CUDA 的后续修复
  （这 312 个提交里包含 `cuda: fixes races in mmid and mmf`、opencl、server、grammar 等）。
  升级代价：必须重跑 §8.5 三条验证（内核代码有变动，不能假设仍是 1195/0）。
- 编译一次约 10~15 分钟（本机全量 HIP 构建实测）。**最终决定：暂不升级**，现有版本已够用。

**网络备忘**（ComfyUI 镜像里拉模型会遇到同样问题）：

- HuggingFace 直连会被代理掐断（`curl: (35) SSL routines::unexpected eof while reading`），
  改用 `hf-mirror.com`（实测 34MB/s）或 `modelscope.cn`。
- GitHub API / codeload 走代理可用；`git`（HTTPS 智能协议）不可用。
- 代理变量已存在：`http_proxy/https_proxy = http://<user>:<token>@10.1.100.10:3120`。

**本次新增的交接物**

| 文件 | 说明 |
|---|---|
| `/root/llama.cpp-master.tar.gz` | 上游最新源码快照（升级备用） |
| `/root/llama.cpp-new/` | 已解压的上游源码（未编译、未打补丁，随时可用） |
| `/root/qwen2.5-0.5b-instruct-q8_0.gguf` | 645MB，Q8_0(**int8**) 小模型；CPU 端已冒烟通过（164 t/s prompt / 26.4 t/s 生成），GPU 端待验证 |

---

## 10. 附：Qwen3.8-27B Q2_K 测试记录（2026-09-12 已完成；两个模型文件当日按要求删除）

> **要复现就直接重新下载**（约 5 分钟，aria2c 8 连接，~35 MiB/s）：
>
> ```bash
> aria2c -x8 -s8 -k1M --file-allocation=none -d /root \
>   "https://cnb.cool/comfyui114514/model/Huihui-Qwen3.8-27B-abliterated-GGUF/-/lfs/c4a5e431ede372af5bf8f57129194bb440d3cf7914d81c63323748ff4fee2ad9?name=Huihui-Qwen3.8-27B-abliterated-Q2_K.gguf"
> aria2c -x8 -s8 -k1M --file-allocation=none -d /root \
>   "https://cnb.cool/comfyui114514/model/Huihui-Qwen3.8-27B-abliterated-GGUF/-/lfs/c9a09064683620bea3d3bfed5d4462e1a97a7d2fff7e5045d6862a0a85eeb5b5?name=mmproj-model-bf16.gguf"
> ```
>
> 删除原因：省镜像体积（11 GiB + 889 MB）。`/root` 现保留：
> `Qwen3.5-4B-Q4_0.gguf`(2.5G)、`qwen3.5-4b-f16.gguf`(7.9G)、`mmproj-BF16.gguf`(645M)、
> `qwen2.5-0.5b-instruct-q8_0.gguf`(645M)。

| 项 | 值 |
|---|---|
| 文本模型 | `/root/Huihui-Qwen3.8-27B-abliterated-Q2_K.gguf`，10 GiB，866 个张量，65 层 —— **已删除**（见上方重新下载命令） |
| 多模态 | `/root/mmproj-model-bf16.gguf`，888 MB —— **已删除** |
| 下载方式 | cnb.cool 的 LFS 链接 + `aria2c -x8 -s8`（约 35 MiB/s，10 GiB 约 5 分钟） |
| 架构 | **`general.architecture = qwen35`** —— 与本机 `Qwen3.5-4B-Q4_0.gguf` **相同** |
| 结论 | **这个 llama.cpp 直接认识该架构，不需要升级**；Q2_K 也在 MMQ 支持列表内 |

注意：`curl -I` 对这个 LFS 链接会返回 **404（误报）**，但 `aria2c` 的 GET 正常，别被 HEAD 结果骗了。

**纯 CPU 冒烟已通过**（`-dev none`，输出正常中文）：

```
> 你好，用一句话介绍你自己
[Start thinking]
用户要求"用一句话介绍自己"，需要简短自介。最终答案应为中文且一句话。无需调用工具，
[ Prompt: 2.2 t/s | Generation: 0.3 t/s ]     ← 27B Q2_K 在 CPU 上的正常速度
```

**GPU 端测试已完成**（清场后，VRAM 归零再跑，全部通过）：

| 测试 | 结果 |
|---|---|
| ① Qwen3.8-27B Q2_K，`-ngl 99` | ✅ 正常中文：`我是一个能帮你解答问题、提供建议并协助完成各种任务的 AI 助手。`<br>`Prompt: 110.2 t/s \| Generation: 28.7 t/s` |
| ② `test-backend-ops -o MUL_MAT -p "q2_K"` | ✅ **11/11 tests passed** |
| ③ Qwen2.5-0.5B **Q8_0（int8）**，`-ngl 99` | ✅ 正常中文：`我是Qwen，由阿里云开发的AI助手`<br>`Prompt: 1209.6 t/s \| Generation: 247.6 t/s` |

对比：同一个 27B Q2_K 在纯 CPU 上是 `2.2 / 0.3 t/s`，**GPU 快了约 95 倍**（生成 28.7 vs 0.3）。
→ 结论：**Qwen3.8-27B 可以直接用，不需要升级 llama.cpp**；Q2_K 与 Q8_0 这两种量化在这个修复下都验证正常。

复现命令：

```bash
cd /root/llama.cpp/build-gfx928/bin && source /opt/dtk/env.sh && export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
./test-backend-ops test -o MUL_MAT -p "q2_K"
./llama-cli -m /root/Huihui-Qwen3.8-27B-abliterated-Q2_K.gguf -ngl 99 -p "你好" -n 64 -t 8 -st --no-warmup < /dev/null
./llama-cli -m /root/qwen2.5-0.5b-instruct-q8_0.gguf -ngl 99 -p "你好" -n 32 -t 8 -st --no-warmup < /dev/null
```

多模态（`--mm /root/mmproj-model-bf16.gguf`）尚未测，需要一张图片样本。

### 10.1 新发现：`-ngl 0` 并不等于「纯 CPU」

这个版本的 llama.cpp 即使 `-ngl 0`，也会**自动把部分算子 offload 到 ROCm 后端**
（栈证据：`ggml_cuda_op_bin_bcast<op_mul>` → `hipLaunchKernel` →
`hsa_executable_freeze` → `BlitKernel::SubmitLinearCopyCommand` →
`InterruptSignal::WaitRelaxed(HSA_WAIT_STATE_ACTIVE)` 在 99.6% CPU 上空转等信号）。

后果：**GPU 卡死时，连"CPU 测试"都会卡在 `Loading model...`**，看起来像 CPU 慢，其实是 GPU 死等。
要真正的纯 CPU，必须加 **`-dev none`**。

（顺带说明：本节的 CPU 数据 2.2 / 0.3 t/s 之所以这么低，是 27B Q2_K 在 CPU 上的正常水平，不是异常。）

---

## 11. 换卡实测：BW（gfx936）— 2026-09-13

### 11.1 设备

```
卡        : BW100, gfx936:sramecc+:xnack- (cc = 0x1000936), 64GB(65520 MiB), wave64, 80 CU, 1500MHz
             （hy-smi/rocminfo 的 Marketing Name 只显示 "BW"；80 CU 这档是 BW100，不是 BW1000）
驱动/DTK  : 同前（hydcu 6.2.28 / DTK 26.04，rocblas 内核库仍只有 gfx906）
```

### 11.2 结论：**源码一行都不用改**

`vendors/hip.h` / `common.cuh` 的 Hygon 列表本来就含 `__gfx936__` 与 `GGML_CUDA_CC_HYGON_GFX936`，
所以 §3 的补丁对 gfx936 同样生效；且 gfx936 **同样没有 mai-insts**
（实测 `clang -x hip --offload-arch=gfx936` 编 mfma 内建函数报 `needs target feature mai-insts`），
`HYGON_DCU_NO_MFMA` 继续保持。

**唯一要做的：按新架构重新编译**（HIP fatbin 不跨架构，gfx928 的二进制在 gfx936 上跑不起来）：

```bash
cd /root/llama.cpp && source /opt/dtk/env.sh
cmake -B build-gfx936 -S . -DCMAKE_BUILD_TYPE=Release -DGGML_HIP=ON \
  -DAMDGPU_TARGETS=gfx936 -DGPU_TARGETS=gfx936 -DGGML_HIP_GRAPHS=ON -DGGML_HIP_NO_VMM=ON \
  -DCMAKE_HIP_COMPILER=/opt/dtk/llvm/bin/clang
cmake --build build-gfx936 -j16 --target llama-cli test-backend-ops llama-bench   # 实测 6m17s（-j16）
```

### 11.3 验证结果

| 项目 | 结果 |
|---|---|
| `test-backend-ops -o MUL_MAT -p q4_0` | 46/46 OK，`2/2 backends passed` |
| `test-backend-ops -o MUL_MAT`（全量） | **1193 OK / 0 FAIL** |
| `llama-cli` Q4_0 中文生成 | 正常，`Prompt 274 t/s \| Generation 114 t/s` |

`llama-bench -ngl 99 -p 512 -n 32 -r 3`（**-fa off，关闭 graph**，原因见 11.4）：

| 模型 | pp512 | tg32 |
|---|---|---|
| Qwen3.5-4B **Q4_0** | **798.05 ± 0.78** | **107.37 ± 0.32** |
| Qwen3.5-4B **F16** | **739.11 ± 1.34** | **77.90 ± 0.13** |
| Qwen2.5-0.5B **Q8_0** | **8534.03 ± 16.19** | **318.42 ± 0.11** |

对比 gfx928（K100_AI，128 CU）：prefill 同量级（740 → 798），生成更快（94.9 → 107）。

### 11.4 新坑：量化 + HIP graph → `hipGraphExecUpdate` 段错误（已定性）

`llama-bench` 跑 tg 阶段必崩（Q4_0 4B，稳定复现）：

```
Launch params (256, 1, 1) are larger than launch bounds (128) for kernel mul_mat_vec_q<Q4_0,1,...>
Segmentation fault (core dumped)
```

- gdb 栈：`hipGraphExecUpdate` ← `ggml_cuda_graph_evaluate_and_capture` ← `llama_decode`，
  崩溃地址在 `/opt/dtk/hip/lib/libgalaxyhip.so.5` 里（**不是 ggml 的代码**）。
- 那条 launch bounds 警告是**误报**：加临时打印实测 host 侧 `cc=0x1000936 table_id=2`，
  枚举里 `MMVQ_PARAMETERS_GCN = 2` → 2 warps = 128 threads，与 device 侧 `CDNA→GCN` 完全一致；
  警告里的 256 是 **grid 维数**（(=256,1,1)）不是 block 维数。
  → **§6.2 的"未定性"到此结案：不是 MMVQ 配置错位，别再去改 `calc_nwarps`。**
- 触发条件：量化（Q4_0）走 MMVQ + HIP graph **更新**路径。~~F16(tg32=80.6) / Q8_0(0.5B, tg32=370.9)
  开 graph 正常~~ → **§14.5 更正：F16 开 graph 同样崩（4/4 稳定复现）**，只有 0.5B Q8_0 侥幸跑通。
  （`llama-cli` 单轮对话开 graph 仍正常。）
- 规避：`export GGML_CUDA_DISABLE_GRAPHS=1`（或构建时 `-DGGML_HIP_GRAPHS=OFF`）。

### 11.5 兼容性：一个二进制覆盖多张卡

`AMDGPU_TARGETS` 可一次给多个架构，编译出的 fatbin 同时能在 K100_AI(gfx928) 与 BW(gfx936) 上跑，
**不需要改任何代码**（代价：每个 .hip 编两遍，编译时间翻倍）：

```bash
-DAMDGPU_TARGETS="gfx928;gfx936" -DGPU_TARGETS="gfx928;gfx936"
```

→ `/root/llama.cpp/build-gfx9xx/bin/`。以后新增卡型只需往这个列表和 §3 的宏列表里加架构名。

### 11.6 gfx936 运行期参数结论（对照《llm插件开发报告》§9.3 的 K100-AI 结论）

| 参数 | K100-AI(gfx928) | **BW100(gfx936) 实测** |
|---|---|---|
| `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` + `HSA_ENABLE_SDMA=0` | ~~必需~~ **不需要**（§13 实测：加上反而 `no ROCm-capable device`） | **不需要**：全程未设也正常出字、无 VMFault |
| flash attention | 必须关（开着 prefill -32%） | ~~建议关~~ → **可开可关**：§14.6 复测 tg 108.3(on) vs 106.6(off)，在噪声内，旧结论不复现 |
| HIP graph | 可用（§13 实测确认，Q4_0 bench 跑通） | 量化模型必崩（§11.4）→ 必须关（**F16 也崩，§14.5**） |
| `--gpu-max-threads-per-block=1024` | ~~需要~~ **本版本无此参数**，传了报 `invalid argument`（§13） | 未加也全过（本 commit 未触发），暂不需要 |

FA 实测（Q4_0，关闭 graph，-r 3，两轮）：

| FA | pp512 | tg32 |
|---|---|---|
| on | 775.95 / 780.25 | 56.88 / 61.68 |
| off | 776.16 / 774.26 | 65.02 / 79.28 |

### 11.7 多架构 fatbin（一份二进制覆盖 K100-AI / BW100 / z100sm）

| 构建目录 | AMDGPU_TARGETS | 状态 |
|---|---|---|
| `build-gfx936` | gfx936 | 单架构，6m17s |
| `build-gfx9xx` | gfx928;gfx936 | K100-AI + BW100，8m50s；gfx936 冒烟 227.2/105.2 t/s 正常 |
| **`build-gfx9-all`** | gfx906;gfx928;gfx936 | 再加 z100sm，11m41s；`strings libggml-hip.so.0` 三个架构各 270 处；gfx936 冒烟 230.1/101.9 t/s，`MUL_MAT -p q4_0` 46/46 通过 |

**z100sm(gfx906) 兼容性判断（已上机实测，结论见 §12）**：cc = `0x1000906` 落在
`GGML_CUDA_CC_IS_GCN` 区间，device 侧 `__gfx906__ → GCN5 → GCN`，与 host 侧天然同表；
上游 `common.cuh`（723 / 770 行）本来就有 `__gfx906__` 特判 → **不需要任何适配**。
且 DTK 的 rocblas Tensile 库里只有 gfx906 内核，z100sm 的 prefill 很可能比 gfx936 更快（待实测）。
⚠️ **不要**把 gfx906 加进 `HYGON_DCU_NO_MFMA` 的宏列表（它不属于 Hygon gfx9 系列）。

> **2026-09-14 更新：以上判断已实测确认，但`build-gfx9-all` 这份 fatbin 在本机跑不起来，
> 必须在本机重编一份 gfx906 的，且跑 bench 必须关 HIP graph。详见 §12。**

---

## 12. 换卡实测（续）：z100sm（gfx906）— 2026-09-14

### 12.1 设备与主机

```
卡        : z100sm, gfx906:sramecc-:xnack- (cc = 0x906), Vega 20, wave64, VRAM 16368 MiB
            （Marketing Name 显示 "Vega 20 [Radeon Pro VII/Radeon Instinct MI50 32GB]"，
              但实际可用只有 16368 MiB，别按 32GB 规划显存）
主机 CPU  : Hygon C86 7185 32-core —— 注意：没有 AVX-512（只有 fma/avx/avx2/f16c/bmi2）
驱动/DTK  : 同前（hydcu / DTK 26.04）
```

### 12.2 结论：源码零改动，但**必须在本机重编**

`§11.7` 的代码判断全部正确：gfx906 走 `vendors/hip.h:179` 的 `__gfx900__/__gfx906__ → GCN5`，
与 host 侧 `GGML_CUDA_CC_IS_GCN` 天然同表，**不需要 §3 的任何补丁，也不需要新增宏**。
已实测 `MUL_MAT` 全量 1193/1193 通过、Q4_0 中文输出正常 → **z100sm 可用**。

但**镜像里现成的 `build-gfx9-all` 在这台机器上一步都跑不了**，两个坑：

#### 坑 A：现成二进制 SIGILL（AVX-512 不匹配）

`build-gfx9-all` 是 `GGML_NATIVE=ON` 在 **C86 7375**（有 AVX-512）上编的，本机是 **C86 7185**（无 AVX-512）→
`libggml-cpu.so` 里全是 `zmm` / `vpbroadcastq %rcx,%xmm0`，进程一起来就 SIGILL：

```
Thread 7 "llama-cli" received signal SIGILL, Illegal instruction.
0x... in ggml_backend_cpu_device_get_extra_buffers_type() from libggml-cpu.so.0
=> vpbroadcastq %rcx,%xmm0        # r64→xmm 形式的 VPBROADCASTQ 是 AVX512 指令
```

**与 DCU 无关**（`-dev none` 纯 CPU 也崩）。修法：重编时关掉 native，显式指定可移植的 AVX2：

```
-DGGML_NATIVE=OFF -DGGML_AVX=ON -DGGML_AVX2=ON -DGGML_FMA=ON -DGGML_F16C=ON
```

#### 坑 B：cmake configure 报 `HSA_HEADER-NOTFOUND`

本容器的 `/opt/hyhal/hsa/include/hsa` 是**空目录**，而 `/opt/dtk/include/hsa` 是指向不存在路径的悬空软链，
导致 `enable_language(HIP)` 的 ABI 检测失败（真正的 `hsa/hsa.h` 在 `/opt/dtk/.hyhal/include`）：

```bash
export CMAKE_PREFIX_PATH=/opt/dtk/.hyhal:$CMAKE_PREFIX_PATH    # configure 前必须加
```

> 注意 `-DHSA_HEADER=...` 不管用：ABI 检测是独立的 try_compile 子项目，不继承这个 cache 变量，必须走环境变量。

### 12.3 构建命令

```bash
cd /root/llama.cpp && source /opt/dtk/env.sh
export CMAKE_PREFIX_PATH=/opt/dtk/.hyhal:$CMAKE_PREFIX_PATH
cmake -B build-gfx906 -S . -DCMAKE_BUILD_TYPE=Release -DGGML_HIP=ON \
  -DAMDGPU_TARGETS=gfx906 -DGPU_TARGETS=gfx906 -DGGML_HIP_GRAPHS=ON -DGGML_HIP_NO_VMM=ON \
  -DGGML_NATIVE=OFF -DGGML_AVX=ON -DGGML_AVX2=ON -DGGML_FMA=ON -DGGML_F16C=ON \
  -DCMAKE_HIP_COMPILER=/opt/dtk/llvm/bin/clang
cmake --build build-gfx906 -j32 --target llama-cli test-backend-ops llama-bench   # 实测 16m49s
```

### 12.4 验证结果

| 项目 | 结果 |
|---|---|
| `test-backend-ops -o MUL_MAT -p q4_0` | ✅ 46/46 OK，`2/2 backends passed` |
| `test-backend-ops -o MUL_MAT`（全量） | ✅ **1193 OK / 0 FAIL**，31 秒跑完 |
| `llama-cli` Qwen3.5-4B **Q4_0** `-ngl 99` | ✅ 正常中文（thinking 过程完整），`Prompt 154.5 \| Gen 70.5 t/s` |
| `llama-cli` Qwen2.5-0.5B **Q8_0** `-ngl 99` | ✅ 正常英文 `Hello! How can I assist you today`，`637.8 / 181.7 t/s` |

### 12.5 性能（`-ngl 99`，**关 graph**，`-p 512 -n 32 -r 3`，`-fa off`）

| 模型 | pp512 | tg32 |
|---|---|---|
| Qwen3.5-4B **Q4_0** | 601.26 ± 0.07 | 70.69 ± 0.26 |
| Qwen3.5-4B **F16** | **1276.11 ± 8.95** | 46.05 ± 0.04 |
| Qwen2.5-0.5B **Q8_0** | 2740.79 ± 0.31 | 201.06 ± 3.37 |

**证实了 §11.7 的预测**：`rocblas` Tensile 库里只有 gfx906 内核（`ls /opt/dtk/lib/rocblas/library | grep gfx` → 178 个全是 gfx906），
所以 F16 prefill **1276 t/s，是 gfx936(739) 的 1.7 倍**——这是三张卡里唯一能吃到调优 GEMM 的。
反过来量化路径（MMQ）不吃 rocblas，gfx906 反而比 gfx936 慢（pp 601 vs 798、tg 70.7 vs 107）。

> `hipblaslt` 库里只有 gfx928 的内核 → `ROCBLAS_USE_HIPBLASLT=1` 在 gfx906 上**没有意义**，不用试。

### 12.6 HIP graph：gfx906 上同样崩，且比 gfx936 更严重

`llama-bench` 不加 `GGML_CUDA_DISABLE_GRAPHS=1` 时，pp512 跑完后进 tg 阶段必崩：

| 模型 | graph ON（默认） |
|---|---|
| Q4_0 4B | ❌ pp512 598.69 后 SIGSEGV |
| **F16 4B** | ❌ pp512 1243.05 后 SIGSEGV（~~gfx936 上 F16 开 graph 是正常的，这里也崩~~ → **§14.5 更正：gfx936 上 F16 同样崩，两者行为一致**） |
| Q8_0 0.5B | ⚠️ 侥幸跑通（215.22 t/s），但别指望 |

栈顶与 §11.4 完全一致，同一个 bug：

```
#0 ... /opt/dtk/hip/lib/libgalaxyhip.so.5
#3 hipGraphExecUpdate () from /opt/dtk/hip/lib/libgalaxyhip.so.5
#4 ggml_cuda_graph_evaluate_and_capture() from libggml-hip.so.0
```

那条 `Launch params (256, 1, 1) are larger than launch bounds (128)` 警告**照旧是误报**（§11.4 已定性，256 是 grid 不是 block）。

- `llama-cli` 单轮/多轮对话**开 graph 正常**（Q4_0 生成 64 token 无异常）→ 崩的是 graph **更新**路径，只有 bench/长会话会撞上。
- 规避：`export GGML_CUDA_DISABLE_GRAPHS=1`（或构建时 `-DGGML_HIP_GRAPHS=OFF`）。

### 12.7 运行期参数结论（补齐 §11.6 的三卡对照）

| 参数 | K100-AI(gfx928) | BW100(gfx936) | **z100sm(gfx906) 实测** |
|---|---|---|---|
| `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` + `HSA_ENABLE_SDMA=0` | ~~必需~~ **不需要**（§13） | 不需要 | **不需要**：全程未设无 VMFault |
| flash attention | 必须关 | 建议关（tg 慢 15~25%） | **无所谓**：pp 600.8 vs 601.3、tg 71.08 vs 70.67，在噪声内 |
| HIP graph | 可用（§13 实测确认） | 量化必崩 | 量化 **和 F16** 都崩 → 必须关 |
| `--gpu-max-threads-per-block` | ~~需要 1024~~ 本版本无此参数（§13） | 不需要 | 未加也全过，不需要 |
| 源码改动 | §3 补丁 | §3 补丁 | **零改动** |

### 12.8 MFMA 探测（顺带把 §8.3 的检查补全）

`/opt/dtk/llvm/bin/clang -x hip --offload-arch=<arch>` 编 `__builtin_amdgcn_mfma_f32_16x16x16f16`：

| arch | 结果 |
|---|---|
| gfx906 | `needs target feature mai-insts` |
| gfx928 / gfx936 | `needs target feature mai-insts`（与 §2.3 一致） |

→ 这台 DTK 的 clang **对 gfx906 也没有 mai-insts**。但因为 gfx906 走 GCN5、`AMD_MFMA_AVAILABLE`
根本不会被定义，ggml 永远不会去编 MFMA 代码 → **不影响构建，也不需要任何处理**。
继续维持 §11.7 的结论：**不要把 gfx906 加进 `HYGON_DCU_NO_MFMA`**。

### 12.9 交接物

| 文件 | 说明 |
|---|---|
| `/root/llama.cpp/build-gfx906/bin/` | **z100sm 上验证过的单架构构建**（llama-cli / test-backend-ops / llama-bench） |
| `/root/llama.cpp/build-gfx9-all2/bin/` | **5 架构 fatbin**，见 §12.10 |
| `/tmp/build906.log`、`/tmp/cfg906.log` | gfx906 构建日志 |
| `/tmp/build_all2.log`、`/tmp/cfg_all2.log` | 5 架构构建日志 |

### 12.10 一份二进制覆盖全部 DCU（2026-09-14 已完成）

`AMDGPU_TARGETS` 一次给全 DTK 默认的 5 个架构（`/opt/dtk/env.sh` 里就是这串），编出来的 fatbin
能同时跑 z100sm(gfx906) / K100_AI(gfx926,gfx928) / BW(gfx936,gfx938)：

```bash
cd /root/llama.cpp && source /opt/dtk/env.sh
export CMAKE_PREFIX_PATH=/opt/dtk/.hyhal:$CMAKE_PREFIX_PATH          # 坑 B，别漏
cmake -B build-gfx9-all2 -S . -DCMAKE_BUILD_TYPE=Release -DGGML_HIP=ON \
  -DAMDGPU_TARGETS="gfx906;gfx926;gfx928;gfx936;gfx938" \
  -DGPU_TARGETS="gfx906;gfx926;gfx928;gfx936;gfx938" \
  -DGGML_HIP_GRAPHS=ON -DGGML_HIP_NO_VMM=ON \
  -DGGML_NATIVE=OFF -DGGML_AVX=ON -DGGML_AVX2=ON -DGGML_FMA=ON -DGGML_F16C=ON \
  -DCMAKE_HIP_COMPILER=/opt/dtk/llvm/bin/clang
cmake --build build-gfx9-all2 -j32 --target llama-cli test-backend-ops llama-bench
```

| 项 | 实测 |
|---|---|
| 编译耗时 | **49m46s**（`-j32`，C86 7185；单架构是 16m49s，5 架构约 3 倍） |
| fatbin 架构 | `strings libggml-hip.so.0 \| grep -oE gfx9[0-9a-z]+` → gfx906 / 926 / 928 / 936 / 938 **各 270 处** |
| `libggml-hip.so.0` 体积 | 21 MiB（单架构约 8 MiB） |
| gfx906 冒烟 | 全量 `MUL_MAT` **1193/1193 / 0 FAIL**；Q4_0 中文正常；`pp512 602.03 / tg32 71.49` |

**多架构没有性能代价**：同一张 gfx906 上，单架构 601.26 / 70.69 vs 5 架构 602.03 / 71.49，在噪声内
（fatbin 只是多了几份 code object，运行时按当前架构挑一份）。

> 这份构建用的是 `GGML_NATIVE=OFF + AVX2`（§12.2 坑 A），所以 **C86 7185 和 7375 两种主机都能跑**，
> 不会再撞 SIGILL —— 旧的 `build-gfx9-all`（AVX-512）只能在那台 7375 上用。

**各卡上机复现命令**（换机器后照抄，三条必做，见 §8.5）：

```bash
cd /root/llama.cpp/build-gfx9-all2/bin && source /opt/dtk/env.sh
export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
export GGML_CUDA_DISABLE_GRAPHS=1              # gfx906/936 必加，gfx928 可不加

./test-backend-ops test -o MUL_MAT -p "q4_0"   # 期望 46/46 OK（gfx928 实测 48/48，见 §13）
./test-backend-ops test -o MUL_MAT             # 期望 0 FAIL（总数因环境而异，见 §13.5）
./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -p "你好" -n 64 -t 8 -st --no-warmup < /dev/null
```

> gfx928 上**不用** `GGML_CUDA_DISABLE_GRAPHS`（§13.4）；`-fa off` 建议加。

> ~~⚠️ gfx926 / gfx928 / gfx936 / gfx938 **只做了编译期覆盖，没有在这台机器上真机验证**（本机只有 gfx906）。~~
> ✅ **2026-09-14 更新：gfx928 已回到 K100_AI 真机验证完毕（§13），gfx936 也已用同一份 fatbin 上机验证（§14）。**
> gfx906 同样在本机实测通过（§12）。仅剩 gfx926 / gfx938 仍是「只编译未上机」。
>
> ⚠️ 旧的「gfx928 额外注意项」（§11.6：需要 `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` + `HSA_ENABLE_SDMA=0`、
> `--gpu-max-threads-per-block=1024`）**已作废，不要照抄** —— 详见 §13.4。

---

## 13. 回到 K100_AI（gfx928）真机复测 — 2026-09-14

> 之前 §1~§6 的记录是在 gfx928 上做的，但 §11/§12 换卡实测时对 gfx928 的运行期参数写了一批
> 「继承自《llm插件开发报告》」的未验证结论。本节是**用 `build-gfx9-all2` 在真 K100_AI 上重新跑一遍**，
> 确认兼容性的同时把那批结论订正。

### 13.1 设备与构建

```
卡        : K100_AI, gfx928:sramecc+:xnack- (0x928), VRAM 65520 MiB, VMM: no, Wave Size: 64
            （rocminfo 里 CPU 侧 Marketing Name 显示 "Hygon C86-4G (OPN:7495)"）
驱动/DTK  : 同前（hydcu 6.2.28 / DTK 26.04）
构建      : /root/llama.cpp/build-gfx9-all2/bin（5 架构 fatbin，GGML_NATIVE=OFF + AVX2）
commit    : 7584430
```

**直接拿现成的 fatbin 跑，不用重编** —— `build-gfx9-all2` 里含 gfx928 的 code object，
且是 AVX2 可移植构建，不存在 §12.2 坑 A 的 SIGILL 问题。

### 13.2 验证结果（全绿）

| 测试 | 结果 |
|---|---|
| `test-backend-ops test -o MUL_MAT -p q4_0` | ✅ **48 OK / 0 FAIL**，`2/2 backends passed` |
| `test-backend-ops test -o MUL_MAT`（全量） | ✅ **1563 项 / 0 FAIL**，`2/2 backends passed` |
| `llama-cli` Qwen3.5-4B **Q4_0** `-ngl 99 -fa off` | ✅ 正常中文，Gen **98.3 t/s** |
| `llama-cli` Qwen2.5-0.5B **Q8_0** `-ngl 99` | ✅ 正常英文，1456.3 / 238.9 t/s |

`llama-bench -ngl 99 -p 512 -n 32 -r 3 -fa off`：

| 模型 | HIP graph | pp512 | tg32 |
|---|---|---|---|
| Qwen3.5-4B **Q4_0** | **on** | **787.63 ± 1.47** | **95.22 ± 0.33** |
| Qwen3.5-4B **Q4_0** | off | 780.08 ± 2.36 | 94.89 ± 0.63 |
| Qwen3.5-4B **F16** | off | 739.59 ± 2.19 | 58.84 ± 0.14 |

（对照历史：§4.3 记录的是 pp512 740.53 / gen 94.9，本次 787.6 / 95.2 —— 同一量级，略好。）

### 13.3 gfx928 可用命令（照抄这份）

```bash
source /opt/dtk/env.sh
cd /root/llama.cpp/build-gfx9-all2/bin
export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
# 不需要 GGML_CUDA_DISABLE_GRAPHS；不要加 UM / SDMA 那两个变量

./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -t 8 -fa off -p "你好" -n 256 -st < /dev/null
```

### 13.4 订正：§11.6 里三条「gfx928 必需项」都是错的

| 旧结论（§11.6 / §12.7） | 实测（§13） |
|---|---|
| 必须 `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` + `HSA_ENABLE_SDMA=0`，否则生成步 `KERNEL VMFault` | **反了**：加上后 `ggml_cuda_init: failed to initialize ROCm: no ROCm-capable device is detected`，一个 token 都跑不了；去掉后全程正常，无 VMFault |
| `--gpu-max-threads-per-block=1024` 需要（DTK 代码对象线程上限 256） | 本 commit **没有这个参数**，传了直接 `error: invalid argument`。该参数属于别的版本/分支 |
| HIP graph「可用」（当时未实测） | 结论对：Q4_0 的 `llama-bench` 开 graph 跑完 pp512 + tg32 无崩溃，性能与关 graph 在噪声内 |

> 那两个 env 变量是在**别的卡/别的驱动**上总结出来的，不要跨环境照抄。
> 现在三张卡（gfx906 / 928 / 936）的统一结论：**只有 gfx906 / gfx936 需要关 HIP graph，
> UM/SDMA 变量三张卡都不需要。**

### 13.5 全量 MUL_MAT 数量变化：1195 → 1563

§4.1 记的是 1195 项，§11/§12 是 1193 项，本次 **1563** 项 —— 不是回归，是 `test-backend-ops`
的用例矩阵随后端能力变化（本次未限制 `-p`，且量化类型集合更全）。
**判据只看 FAIL 数：0 即通过**，不要跨环境对比总数。

### 13.6 交接物

| 文件 | 说明 |
|---|---|
| `/tmp/gfx928_test.sh`、`/tmp/gfx928_test.log` | 第一轮（带 UM/SDMA 变量，暴露 §13.4 的问题） |
| `/tmp/gfx928_test2.sh`、`/tmp/gfx928_test2.log` | **第二轮，最终结论**，子日志 `/tmp/t1..t6.log` |

> 日志在 `/tmp`，容器重启会丢；要留档就复制到 `/root/CodeBuddy/` 下。

---

## 14. BW100(gfx936) 复测：`build-gfx9-all2` 上机 — 2026-09-14

> §12.10 那份 5 架构 fatbin 当时只验证了 gfx906 / gfx928，**gfx936 属于「只编译未上机」**。
> 本节回到 BW100 把它补齐，顺带订正 §11.4 / §11.6 里两条被推翻的旧结论。

### 14.1 设备与构建

```
卡        : gfx936（rocminfo: Name gfx936 / Marketing "BW"），hy-smi 起始 VRAM 0%
主机 CPU  : Hygon C86，flags 含 avx512f → 旧的 build-gfx9-all（AVX-512）在这台也能跑，
            但本次统一用可移植的那份
构建      : /root/llama.cpp/build-gfx9-all2/bin（5 架构 fatbin，GGML_NATIVE=OFF + AVX2）
commit    : 7584430
```

### 14.2 算子正确性（全绿）

| 测试 | 结果 |
|---|---|
| `test-backend-ops test -o MUL_MAT -p q4_0` | ✅ **46/46**，2/2 backends passed |
| `test-backend-ops test -o MUL_MAT`（全量） | ✅ **1193/1193 / 0 FAIL** |
| `-p q2_K` | ✅ 11/11 |
| `-p q8_0` | ✅ 47/47 |
| `-p "iq"` | ✅ 132/132 |

（全量 1193 与 §11.3 一致；§13.5 的「只看 FAIL 数」判据继续成立。）

### 14.3 端到端生成（`GGML_CUDA_DISABLE_GRAPHS=1`）

| 模型 | 结果 |
|---|---|
| Qwen3.5-4B **Q4_0** `-ngl 99 -fa off` | ✅ 中文正常（thinking 过程完整），Gen **100.8 t/s** |
| Qwen3.5-4B **F16** `-ngl 99 -fa off` | ✅ 中文正常，Gen **74.1 t/s** |
| Qwen2.5-0.5B **Q8_0** `-ngl 99` | ✅ `Hello! How can I assist you today?`，1794.3 / **284.0 t/s** |

### 14.4 性能（`-ngl 99 -p 512 -n 32 -r 3 -fa off`，**关 graph**）

| 模型 | pp512 | tg32 |
|---|---|---|
| Qwen3.5-4B **Q4_0** | **794.84 ± 0.61** | **105.42 ± 0.10** |
| Qwen3.5-4B **F16** | **733.28 ± 1.03** | **76.96 ± 0.14** |
| Qwen2.5-0.5B **Q8_0** | **8484.38 ± 11.62** | **285.78 ± 1.83** |

对照 §11.3 的单架构 `build-gfx936`（798.05/107.37、739.11/77.90、8534.03/318.42）
→ 同一量级。**fatbin 无性能代价**：Q8_0 用 `-r 5` 重测，
`build-gfx9-all2` **287.93 ± 1.34** vs `build-gfx936` **287.96 ± 0.40**，逐位一致。

### 14.5 更正 §11.4：F16 开 HIP graph **也崩**（不是只有量化）

graph 开启（不设 `GGML_CUDA_DISABLE_GRAPHS`）跑 `llama-bench`，看真实退出码：

| 模型 | graph=ON | 退出码 |
|---|---|---|
| Qwen3.5-4B **Q4_0** | pp512 769.30 后崩 | **139 (SIGSEGV)** |
| Qwen3.5-4B **F16** | pp512 712.42 后崩 | **139 (SIGSEGV)** |
| Qwen2.5-0.5B **Q8_0** | 侥幸跑完，pp 8350 / tg 337.31 | 0 |

- F16 的崩溃 **4/4 稳定复现**，且 `build-gfx9-all2` 与 `build-gfx936` 两种构建都崩
  → 不是 fatbin 引入的，是 gfx936 自身行为。
- 栈顶/现象与 §11.4 一致（`hipGraphExecUpdate` ← `ggml_cuda_graph_evaluate_and_capture`，
  外加那条**照旧是误报**的 `Launch params (256, 1, 1) ... launch bounds (128)`）。
- → §11.4 里「F16 开 graph 正常，tg32=80.6」**作废**。gfx936 与 gfx906 行为一致：
  **量化和 F16 都要关 graph**，只有 0.5B 那种小模型可能蒙混过关，别指望。

### 14.6 更正 §11.6：flash attention 开关在噪声内

Q4_0，`-r 3`，两轮交叉：

| 轮次 | -fa | pp512 | tg32 |
|---|---|---|---|
| 1 | on | 786.69 ± 1.07 | **108.31 ± 0.21** |
| 1 | off | 781.54 ± 1.63 | 106.57 ± 0.20 |
| 2 | on | 781.03 ± 2.00 | **107.88 ± 0.25** |
| 2 | off | 769.27 ± 0.50 | 106.55 ± 0.15 |

→ 与 §11.6「FA 让 tg 慢 15~25%」**不符**；本次 FA **on 反而快约 1.5%**，pp 也在噪声内。
结论：**gfx936 上 FA 可开可关，不用刻意关**（旧结论别再照抄；§11.6 当时的 tg 绝对值只有 56~79，
怀疑那批数据受了环境/graph 状态干扰）。

### 14.7 BW100 当前推荐命令

```bash
source /opt/dtk/env.sh
cd /root/llama.cpp/build-gfx9-all2/bin
export LD_LIBRARY_PATH=$PWD:$LD_LIBRARY_PATH
export GGML_CUDA_DISABLE_GRAPHS=1          # 必加（§14.5）
# 不要加 GGML_CUDA_ENABLE_UNIFIED_MEMORY / HSA_ENABLE_SDMA

./llama-cli -m /root/Qwen3.5-4B-Q4_0.gguf -ngl 99 -t 8 -p "你好" -n 256 -st < /dev/null
```

> `llama-cli` 单轮对话开 graph 也能跑，但 bench / 长会话必崩 → **统一关掉最省心**。
> FA 不必刻意关（§14.6）。

### 14.8 交接物

| 文件 | 说明 |
|---|---|
| `/tmp/bw100_retest.sh`、`/tmp/bw100_retest.log` | 算子正确性（q4_0 / 全量 / q2_K / q8_0 / iq） |
| `/tmp/bw100_gen.sh`、`/tmp/bw100_gen.log` | 三种模型端到端生成 |
| `/tmp/bw100_bench.sh`、`/tmp/bw100_bench.log` | graph on/off × 3 模型的 bench |
| `/tmp/bw100_graph.sh`、`/tmp/graphon_*.log` | graph=ON 的真实退出码（139） |
| `/tmp/bw100_f16graph.sh`、`/tmp/f16_build-*_*.log` | F16 崩溃在两种构建上的复现 |
| `/tmp/bw100_cmp.sh` | fatbin vs 单架构、FA on/off 对照 |

> 日志在 `/tmp`，容器重启会丢；要留档就复制到 `/root/CodeBuddy/` 下。
