# kitchen hip 后端 gfx928 构建尝试（2026-08-31，编译期实锤封死）

背景：用户问 ModelAttentionBackend 没识别到 kitchen 选项，拍板「先装，装完有问题再测试，帮别人踩坑」。此前 skill 只有"不值得编"的判断，本次按用户指示做了完整构建尝试，拿到指令集层面的硬失败证据。

## 结论（一句话）
kitchen hip 后端的 WMMA kernel 在 gfx928（CDNA/MFMA 指令集）上编译期直接失败，8 处错误全是 `needs target feature gfx11-insts`/`gfx10-insts`。移植 = 重写 mma.h + 全部 attention/GEMM kernel 为 MFMA 版，非补丁可解。**终态，不要再重试编译。**

## 尝试路径与命令（可直接复用）

### 1. 前置探针（最低成本判断，5 分钟）
```bash
# WMMA builtin 在 gfx928 上认不认
cat > /tmp/wmma_probe.hip << 'EOF'
#include <hip/hip_runtime.h>
#include <cstdint>
typedef int v8i __attribute__((ext_vector_type(8)));
typedef int v16i __attribute__((ext_vector_type(16)));
typedef unsigned char v16u __attribute__((ext_vector_type(16)));
extern "C" __global__ void probe(v16u a, v16u b, v8i c, v8i* out) {
    v8i r = __builtin_amdgcn_wmma_i32_16x16x16_iu8_w32(true, a, true, b, c, false);
    out[threadIdx.x] = r;
}
EOF
cd /tmp && hipcc --offload-arch=gfx928 -c wmma_probe.hip -o wmma_probe.o
# → error: '__builtin_amdgcn_wmma_i32_16x16x16_iu8_w32' needs target feature gfx11-insts
```
这一步失败即基本宣判（后续全量构建只是把失败点摆全）。

### 2. 全量构建（隔离副本，不碰 site-packages/运行实例）
```bash
rm -rf /tmp/kitchen_gfx928 && cp -a /public/home/accma5g3vg/kitchen_src /tmp/kitchen_gfx928
rm -rf /tmp/kitchen_gfx928/build /tmp/kitchen_gfx928/comfy_kitchen.egg-info
# patch architectures.json: wmma_gfx11 组追加 "gfx928"
# setup.py 路线（COMFY_HIP_ARCHS=gfx928 + build_ext --hip）：死在 CMake 4.4 拒绝 hipcc wrapper：
#   "CMAKE_HIP_COMPILER is set to the hipcc wrapper ... This is not supported. Use Clang directly"
# 绕过 setup.py 直接 cmake（关键：用真 clang++）：
cmake -S /tmp/kitchen_gfx928/comfy_kitchen/backends/hip -B /tmp/kitchen_gfx928/build/hip -G Ninja \
  -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/tmp/kitchen_gfx928/comfy_kitchen/backends/hip \
  -DCMAKE_BUILD_TYPE=Release -DPython_EXECUTABLE=/public/home/accma5g3vg/venv/bin/python \
  -DCOMFY_ENABLE_LINEINFO=OFF \
  -DCMAKE_HIP_COMPILER=/opt/dtk/llvm/bin/clang++ \
  -DCMAKE_CXX_COMPILER=/opt/dtk/llvm/bin/clang++ \
  -DCMAKE_PREFIX_PATH=/opt/dtk -DCMAKE_HIP_COMPILER_ROCM_ROOT=/opt/dtk \
  -DCOMFY_HIP_ARCHS=gfx928
# configure 通过（7.5s，gfx928 被接受）
cmake --build /tmp/kitchen_gfx928/build/hip --parallel 6   # → 编译失败
```

### 3. 失败错误清单（ninja 全量，8 处 sort -u）
```
mma.h:265  __builtin_amdgcn_wmma_f32_16x16x16_bf16_w32   needs target feature gfx11-insts
mma.h:299  __builtin_amdgcn_wmma_i32_16x16x16_iu8_w32    needs target feature gfx11-insts
mma.h:320  __builtin_amdgcn_wmma_i32_16x16x16_iu4_w32    needs target feature gfx11-insts
mma.h:338  __builtin_amdgcn_wmma_f32_16x16x16_bf16_w32   needs target feature gfx11-insts
mma.h:351  __builtin_amdgcn_wmma_f32_16x16x16_f16_w32    needs target feature gfx11-insts
sage_common.h:115  __builtin_amdgcn_permlanex16          needs target feature gfx10-insts
（sage_attention/ 路径下 iu8 同类错误 2 处）
```
文件内证据：`/tmp/kitchen_gfx928`（/tmp 不进镜像，重启丢，本文件是持久记录）。

## 关键事实
- 上游 setup.py `setup_hip_extension()` 注释原文：CDNA "uses MFMA rather than WMMA … has no path through these sources at all" —— 上游明确不支持 CDNA，不是漏配置。
- hip 后端架构白名单（architectures.json）只有 RDNA2 elementwise（gfx1030-1036）+ RDNA3/4 WMMA（gfx11xx/gfx12xx），无任何 CDNA 条目。
- MFMA builtin 形态不同（`__builtin_amdgcn_mfma_i32_16x16x32_i8` 16 列累加器 vs WMMA 8 列、参数是 long 簧位标志），且 int8_attn.hip 的在线 softmax 算法依赖 WMMA "S^T 转置、每 lane 一列" 的 fragment 排布 —— 算法与指令布局耦合，不是换个 builtin 就能编。
- 机器规格不是瓶颈：115GB RAM / 6 核，旧结论"4GB cgroup OOM 风险"作废（那是另一台 4GB 无卡机的教训，别再套用）。
- 另一探针副产品：DTK 26.04 的 clang 连 gfx1030 的 ROCm device library 都没有（`cannot find ROCm device library for gfx1030`），只带 gfx9xx —— DTK 环境想编 RDNA 目标也不可行。

## 后续若要推进的唯一路径
等 Hygon 官方/上游出 CDNA(MFMA) kernel 支持或 DTK 提供 ROCm 上游 gfx928 支持；届时重跑本文件探针（1 分钟）即可判断，无需再走全量构建。
